system/lib/wasmfs/file_table.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the open file table of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "file_table.h"
+
+namespace wasmfs {
+
+std::vector<std::shared_ptr<OpenFileState>> FileTable::entries;
+
+static __wasi_errno_t writeStdBuffer(const uint8_t* buf, __wasi_size_t len,
+  void (*console_write)(const char*), std::vector<char>& fd_write_buffer) {
+  for (__wasi_size_t j = 0; j < len; j++) {
+    uint8_t current = buf[j];
+    if (current == '\0' || current == '\n') {
+      fd_write_buffer.push_back('\0'); // for null-terminated C strings
+      console_write(&fd_write_buffer[0]);
+      fd_write_buffer.clear();
+    } else {
+      fd_write_buffer.push_back(current);
+    }
+  }
+  return __WASI_ERRNO_SUCCESS;
+}
+
+class StdinFile : public File {
+
+  __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) override {
+    return __WASI_ERRNO_INVAL;
+  }
+
+  __wasi_errno_t read(const __wasi_iovec_t* iovs, size_t iovs_len, __wasi_size_t* nread) override {
+    emscripten_console_log("StdinFile::read() has not been implemented yet.");
+    return __WASI_ERRNO_INVAL;
+  };
+
+public:
+  File::Handle get() { return Handle(*this); }
```

Comment:
```suggestion
```
Actually, since this does the same thing as `File::get()`, I don't think we need a new definition here. or in other `File` subclasses.